### ISSUE TITLE
Reloading weights give small difference

### ISSUE THREAD

**Issue by Shiro-LK** [tensorflow/tensorflow#46645]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, I modified a TPU colab notebook
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: None
- TensorFlow installed from (source or binary): colab
- TensorFlow version (use command below): 2.4
- Python version:
- Bazel version (if compiling from source): None
- GCC/Compiler version (if compiling from source): None
- CUDA/cuDNN version: TPU v2
- GPU model and memory: None

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
During the training, I am computing the validation accuracy at each epochs, with a callback function and then save the weight when the model improved.
When reloading the weight inside the same session, the model give small difference regarding the validation metric. I don't know if the issue comes from the loading or saving part.

**Describe the expected behavior**
the model should has the same validation performance .

**Standalone code to reproduce the issue**

link of the colab to reproduce the results : 
https://colab.research.google.com/drive/1LnEgPmVtfuX9yaFZftpbOUlPPU4ecUvL#scrollTo=7Qv8rC4aVOFB&uniqifier=2

As you can see : 0.9588235294117647 is different from  0.9573529411764706
> Epoch 1/5
> 
> Epoch 00001: LearningRateScheduler reducing learning rate to 1e-05.
>  6/23 [======>.......................] - ETA: 6s - loss: 1.6750 - accuracy: 0.2327WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0099s vs `on_train_batch_end` time: 11.4604s). Check your callbacks.
> WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0099s vs `on_train_batch_end` time: 11.4604s). Check your callbacks.
> 23/23 [==============================] - 131s 410ms/step - loss: 1.6348 - accuracy: 0.2524
> val_loss: 1.5019999742507935 -- val_categorical_accuracy: 0.365 -- samples : 680
> 
>  model saved, improvement from 0.0 to 0.36470588235294116
> 
> Epoch 2/5
> 
> Epoch 00002: LearningRateScheduler reducing learning rate to 8.8e-05.
> 23/23 [==============================] - 9s 410ms/step - loss: 1.1962 - accuracy: 0.5953
> val_loss: 0.5659999847412109 -- val_categorical_accuracy: 0.854 -- samples : 680
> 
>  model saved, improvement from 0.36470588235294116 to 0.8544117647058823
> 
> Epoch 3/5
> 
> Epoch 00003: LearningRateScheduler reducing learning rate to 0.000166.
> 23/23 [==============================] - 9s 411ms/step - loss: 0.4679 - accuracy: 0.8442
> val_loss: 0.19499999284744263 -- val_categorical_accuracy: 0.934 -- samples : 680
> 
>  model saved, improvement from 0.8544117647058823 to 0.9338235294117647
> 
> Epoch 4/5
> 
> Epoch 00004: LearningRateScheduler reducing learning rate to 0.000244.
> 23/23 [==============================] - 9s 412ms/step - loss: 0.2451 - accuracy: 0.9188
> val_loss: 0.125 -- val_categorical_accuracy: 0.954 -- samples : 680
> 
>  model saved, improvement from 0.9338235294117647 to 0.9544117647058824
> 
> Epoch 5/5
> 
> Epoch 00005: LearningRateScheduler reducing learning rate to 0.000322.
> 23/23 [==============================] - 10s 416ms/step - loss: 0.1769 - accuracy: 0.9334
> val_loss: 0.1289999932050705 -- val_categorical_accuracy: 0.959 -- samples : 680
> 
>  model saved, improvement from 0.9544117647058824 to 0.9588235294117647
> 
> Loading best model... ckpt_model.h5
> final acc :  0.9573529411764706

**Comment by kashyapraval** [comment]
Unable to reproduce issue:

Downloading data from https://github.com/qubvel/efficientnet/releases/download/v0.0.1/efficientnet-b5_noisy-student_notop.h5
115261440/115255328 [==============================] - 2s 0us/step
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
efficientnet-b5 (Functional) (None, 11, 11, 2048)      28513520  
_________________________________________________________________
global_average_pooling2d (Gl (None, 2048)              0         
_________________________________________________________________
dense (Dense)                (None, 5)                 10245     
=================================================================
Total params: 28,523,765
Trainable params: 28,351,029
Non-trainable params: 172,736
_________________________________________________________________
Epoch 1/5

Epoch 00001: LearningRateScheduler reducing learning rate to 1e-05.
 6/23 [======>.......................] - ETA: 6s - loss: 1.6350 - accuracy: 0.2264WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0103s vs `on_train_batch_end` time: 11.4859s). Check your callbacks.
WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0103s vs `on_train_batch_end` time: 11.4859s). Check your callbacks.
23/23 [==============================] - 137s 411ms/step - loss: 1.6074 - accuracy: 0.2476
val_loss: 1.5019999742507935 -- val_categorical_accuracy: 0.384 -- samples : 680

 model saved, improvement from 0.0 to 0.38382352941176473

Epoch 2/5

Epoch 00002: LearningRateScheduler reducing learning rate to 8.8e-05.
23/23 [==============================] - 10s 414ms/step - loss: 1.1882 - accuracy: 0.6374
val_loss: 0.5320000052452087 -- val_categorical_accuracy: 0.851 -- samples : 680

 model saved, improvement from 0.38382352941176473 to 0.8514705882352941

Epoch 3/5

Epoch 00003: LearningRateScheduler reducing learning rate to 0.000166.
23/23 [==============================] - 10s 415ms/step - loss: 0.4235 - accuracy: 0.8747
val_loss: 0.17399999499320984 -- val_categorical_accuracy: 0.934 -- samples : 680

 model saved, improvement from 0.8514705882352941 to 0.9338235294117647

Epoch 4/5

Epoch 00004: LearningRateScheduler reducing learning rate to 0.000244.
23/23 [==============================] - 10s 415ms/step - loss: 0.2045 - accuracy: 0.9396
val_loss: 0.12200000137090683 -- val_categorical_accuracy: 0.954 -- samples : 680

 model saved, improvement from 0.9338235294117647 to 0.9544117647058824

Epoch 5/5

Epoch 00005: LearningRateScheduler reducing learning rate to 0.000322.
23/23 [==============================] - 10s 426ms/step - loss: 0.1671 - accuracy: 0.9411
val_loss: 0.19900000095367432 -- val_categorical_accuracy: 0.932 -- samples : 680
Loading best model... ckpt_model.h5
final acc :  0.9544117647058824

**Comment by Shiro-LK** [comment]
@kashyapraval  I thank you for your reply. I have re-check it few times again few minutes ago and I still have the issue with tpu, but not gpu. Were you using TPU ?

**Comment by kashyapraval** [comment]
@Shiro-LK after some trials, yes able to reproduce the issue. Shall have a look at it.

**Comment by amahendrakar** [comment]
@Shiro-LK,
I did not face any issues while running the code with TF v2.4 on TPUs. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/5894855508d0bd676e7d6b78a221fa17/46645.ipynb#scrollTo=5wU96EmMN6f-&line=1&uniqifier=1). Thanks!

**Comment by Shiro-LK** [comment]
@amahendrakar  I thank you for your reply. Can you launch few times the script ? I try again, and it seems to have some randomness. It will not bug 100% of the time which is very weird.

**Comment by kashyapraval** [comment]
Even when the accuracy scores match, the prediction scores do not match exactly. Two checks made:
1. Recording prediction scores during callbacks(end of each epoch) and comparing to the scores after loading corresponding to the checkpoint model. There is a slight mismatch (~always: even when the accuracies match)
2. Using tpu_strategy.scope() Vs otherwise: Last cell in the [gist](https://colab.research.google.com/drive/1LnEgPmVtfuX9yaFZftpbOUlPPU4ecUvL?usp=sharing) 
The prediction scores still differ slightly. Suspected precision error somewhere?

**Comment by Shiro-LK** [comment]
@kashyapraval  thank you for your reply. I will double check if the issue is still here with gpu to be sure it is related only to tpu

**Comment by kashyapraval** [comment]
@Shiro-LK the issue is with the TPU. [Here](https://colab.research.google.com/drive/15OZHS6YX5ZpBVGS8Mru1WKae91WdiXZz?usp=sharing) you may observe that irrespective of what pretrained model we choose, the difference in weights crops up (in mostly in the conv layers
Moreover the difference exists when exactly one is loaded in the TPUStrategy's scope.

**Comment by Shiro-LK** [comment]
@kashyapraval  Thanks, you are right.

**Comment by ymodak** [comment]
Can you try setting `seed` for the `kernel_initializer` in all the `tf.keras.layers` used in your model?
See similar [issue](https://github.com/tensorflow/tensorflow/issues/47722#issuecomment-806146182) to know more.
